bin/node/cli/tests/websocket_server.rs

```diff
@@ -0,0 +1,423 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use async_std::net::{TcpListener, TcpStream};
+use core::{fmt, pin::Pin, str};
+use futures::{channel::mpsc, prelude::*};
+use soketto::handshake::{server::Response, Server};
+use std::{collections::HashMap, io, net::SocketAddr};
+
+/// Configuration for a [`WsServer`].
+pub struct Config {
+	/// IP address to try to bind to.
+	pub bind_address: SocketAddr,
+
+	/// Maximum size, in bytes, of a frame sent by the remote.
+	///
+	/// Since the messages are entirely buffered before being returned, a maximum value is
+	/// necessary in order to prevent malicious clients from sending huge frames that would
+	/// occupy a lot of memory.
+	pub max_frame_size: usize,
+
+	/// Number of pending messages to buffer up for sending before the socket is considered
+	/// unresponsive.
+	pub send_buffer_len: usize,
+
+	/// Pre-allocated capacity for the list of connections.
+	pub capacity: usize,
+}
+
+/// Identifier for a connection with regard to a [`WsServer`].
+///
+/// After a connection has been closed, its [`ConnectionId`] might be reused.
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct ConnectionId(u64);
+
+/// WebSockets listening socket and list of open connections.
+pub struct WsServer {
```

Comment:
I still think that all this code is just a huge overkill, as we don't require connections support, being notified about new connections or whatever.

In the end we just want to have Strings that are being send.

However, to continue here, I will approve it...

Reply:
Ok I didn't understand that. I simplified the code

Reply:
@niklasad1 @maciejhirsz Don't we have something nimbler we can use for this?

Reply:
yeah, in `jsonrpsee` but it's not published on crates.io so I don't think it can be used here.

However, it just responds with a hardcoded response/subscription this seems much more complicated and I haven't read the tests.

Reply:
There is this: https://crates.io/crates/embedded-websocket